File: src/main/java/com/footverse/order/dto/OrderDetailResponse.java
package com.footverse.order.dto;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

import com.footverse.order.entity.OrderStatus;
import com.footverse.order.entity.PaymentMethod;
import com.footverse.order.entity.PaymentStatus;

/**
 * The full detail of an order with its checkout snapshots (dto-spec §15). Money and shipping fields
 * are frozen at checkout and returned exactly as persisted. The {@code items} list is an aggregate
 * assembled by the service from the order's lines.
 *
 * @param id                     the order id
 * @param orderCode              the immutable order code
 * @param status                 the current order status
 * @param paymentMethod          the payment method ({@code COD})
 * @param paymentStatus          the current payment status
 * @param subtotal               the snapshotted subtotal
 * @param discountAmount         the snapshotted discount (0 when no coupon)
 * @param shippingFee            the snapshotted shipping fee
 * @param total                  the snapshotted total
 * @param couponCode             the applied coupon code, if any
 * @param shippingRecipientName  the shipping recipient name snapshot
 * @param shippingRecipientPhone the shipping recipient phone snapshot
 * @param shippingProvince       the shipping province snapshot
 * @param shippingDistrict       the shipping district snapshot
 * @param shippingWard           the shipping ward snapshot
 * @param shippingStreetAddress  the shipping street address snapshot
 * @param note                   the customer note, if any
 * @param items                  the order lines
 * @param createdAt              the creation timestamp
 * @param cancelledAt            the cancellation timestamp, if cancelled
 * @param deliveredAt            the delivery timestamp, if delivered
 */
public record OrderDetailResponse(
        Long id,
        String orderCode,
        OrderStatus status,
        PaymentMethod paymentMethod,
        PaymentStatus paymentStatus,
        BigDecimal subtotal,
        BigDecimal discountAmount,
        BigDecimal shippingFee,
        BigDecimal total,
        String couponCode,
        String shippingRecipientName,
        String shippingRecipientPhone,
        String shippingProvince,
        String shippingDistrict,
        String shippingWard,
        String shippingStreetAddress,
        String note,
        List<OrderItemResponse> items,
        LocalDateTime createdAt,
        LocalDateTime cancelledAt,
        LocalDateTime deliveredAt) {
}

File: app/lib/features/order/models/order_detail_response.dart
import 'package:json_annotation/json_annotation.dart';

import '../../product/models/decimal_json.dart';
import 'order_item_response.dart';
import 'order_status.dart';
import 'payment_method.dart';
import 'payment_status.dart';

part 'order_detail_response.g.dart';

/// The full detail of an order with its checkout snapshots (dto-spec §15).
///
/// Money and shipping fields are frozen at checkout and returned exactly as
/// persisted; the client renders them and never recomputes them (dto-spec §1;
/// business-rules → Order). [couponCode], [note], [cancelledAt], and
/// [deliveredAt] are present only when they apply.
@JsonSerializable(createToJson: false)
class OrderDetailResponse {
  const OrderDetailResponse({
    required this.id,
    required this.orderCode,
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.subtotal,
    required this.discountAmount,
    required this.shippingFee,
    required this.total,
    required this.shippingRecipientName,
    required this.shippingRecipientPhone,
    required this.shippingProvince,
    required this.shippingDistrict,
    required this.shippingWard,
    required this.shippingStreetAddress,
    required this.items,
    required this.createdAt,
    this.couponCode,
    this.note,
    this.cancelledAt,
    this.deliveredAt,
  });

  factory OrderDetailResponse.fromJson(Map<String, dynamic> json) =>
      _$OrderDetailResponseFromJson(json);

  final int id;
  final String orderCode;
  final OrderStatus status;
  final PaymentMethod paymentMethod;
  final PaymentStatus paymentStatus;
  @JsonKey(fromJson: decimalFromJson)
  final double subtotal;
  @JsonKey(fromJson: decimalFromJson)
  final double discountAmount;
  @JsonKey(fromJson: decimalFromJson)
  final double shippingFee;
  @JsonKey(fromJson: decimalFromJson)
  final double total;
  final String? couponCode;
  final String shippingRecipientName;
  final String shippingRecipientPhone;
  final String shippingProvince;
  final String shippingDistrict;
  final String shippingWard;
  final String shippingStreetAddress;
  final String? note;
  final List<OrderItemResponse> items;
  final DateTime createdAt;
  final DateTime? cancelledAt;
  final DateTime? deliveredAt;
}
